Source/WebCore/Modules/mediasource/MediaSource.cpp

```diff
@@ -245,56 +247,33 @@ void MediaSource::completeSeek()
     auto seekTarget = *m_pendingSeekTarget;
     m_pendingSeekTarget.reset();
 
-    struct SeeksCallbackAggregator final : public RefCounted<SeeksCallbackAggregator> {
-        SeeksCallbackAggregator(MediaTime target, MediaSource& source, CompletionHandler<void(const MediaTime&)>&& completionHandler)
-            : time(target)
-            , mediaSource(source)
-            , completionHandler(WTFMove(completionHandler))
-        {
-            ASSERT(this->completionHandler);
+    Ref<MediaTimePromise> promise = SourceBuffer::ComputeSeekPromise::all(RunLoop::main(), WTF::map(*m_activeSourceBuffers, [&](auto&& sourceBuffer) {
+        return sourceBuffer->computeSeekTime(seekTarget);
+    }))->whenSettled(RunLoop::main(), [time = seekTarget.time, protectedThis = Ref { *this }] (auto&& results) mutable {
+        if (!results)
+            return MediaTimePromise::createAndReject(-1);
+        auto seekTime = time;
+        for (auto& result : *results) {
+            if (abs(time - result) > abs(time - seekTime))
+                seekTime = result;
         }
 
-        ~SeeksCallbackAggregator()
-        {
-            auto seekTime = time;
-            for (auto& result : seekResults) {
-                if (result.isInvalid()) {
-                    completionHandler(MediaTime::invalidTime());
-                    return;
-                }
-                if (abs(time - result) > abs(time - seekTime))
-                    seekTime = result;
-            }
-            completionHandler(seekTime);
-
-            // 4. Resume the seek algorithm at the "Await a stable state" step.
-            mediaSource->monitorSourceBuffers();
-        }
+        // 4. Resume the seek algorithm at the "Await a stable state" step.
+        protectedThis->monitorSourceBuffers();
 
-        MediaTime time;
-        Ref<MediaSource> mediaSource;
-        CompletionHandler<void(const MediaTime&)> completionHandler;
-        Vector<MediaTime> seekResults;
-    };
-
-    auto callbackAggregator = adoptRef(*new SeeksCallbackAggregator(seekTarget.time, *this, WTFMove(m_seekCompletedHandler)));
-
-    for (auto& sourceBuffer : *m_activeSourceBuffers) {
-        sourceBuffer->computeSeekTime(seekTarget, [callbackAggregator](const MediaTime& seekTime) {
-            callbackAggregator->seekResults.append(seekTime);
-        });
-    }
+        return MediaTimePromise::createAndResolve(seekTime);
+    });
+    promise->chainTo(WTFMove(*m_seekTargetPromise));
+    m_seekTargetPromise.reset();
 }
 
-void MediaSource::seekToTime(const MediaTime& time, CompletionHandler<void()>&& completionHandler)
+Ref<GenericPromise> MediaSource::seekToTime(const MediaTime& time)
 {
-    if (isClosed()) {
-        completionHandler();
-        return;
-    }
+    if (isClosed())
+        return GenericPromise::createAndReject(-1);
```

Comment:
This pattern is repeated a lot; I assume that `-1` will be replaced with a `PlatformMediaError` in a future patch?

Reply:
Yes... The "PR#4" adds PlatformMediaError and those are replaced everywhere